### PR TITLE
gh-111924: Use PyMutex for runtime global locks.

### PR DESCRIPTION
**Pull request by colesbury**
This replaces some usages of `PyThread_type_lock` with `PyMutex`, which does not require memory allocation to initialize.

This simplifies some of the runtime initialization and is also one step towards avoiding changing the default raw memory allocator during initialize/finalization, which can be non-thread-safe in some circumstances.

* Issue: gh-111924

### CODE REVIEW

**Comment by colesbury** [comment]
@ericsnowcurrently, when you get a chance, would you please review this?

`PyMutex_Lock()` behaves similarly to `PyThread_acquire_lock()` except that `PyMutex_Lock()` will temporarily release the GIL if it needs to block. That's mostly fine, but I think not what we want for the `HEAD_LOCK()` macro, so that explicitly passes a flag to *not* release the GIL (or detach in `--disable-gil` builds).

File: Python/import.c
Comment (ericsnowcurrently) [inline Python/import.c:418]: nice!

File: Python/pylifecycle.c
Comment (ericsnowcurrently) [inline Python/pylifecycle.c:3058]: It may not make much difference, but I'm pretty sure `Py_AtExit()` is only ever called at times when there shouldn't be any other Python-attached threads (e.g. during shutdown), so it might be worth checking.  If that's true then we could do `PyMutex_LockFlags(&state->mutex, _Py_LOCK_DONT_DETACH)` here.

File: Python/ceval_gil.c
Comment (ericsnowcurrently) [inline Python/ceval_gil.c]: It might be nice if we could have an equivalent here:

```c
    _PyMutex_at_fork_reinit(&pending->mutex);
```

File: Python/crossinterp.c
Comment (ericsnowcurrently) [inline Python/crossinterp.c:459]:
```suggestion
    if (registry->global) {
        PyMutex_Lock(&registry->mutex);
    }
    // else: Within an interpreter we rely on the GIL instead of a separate lock.
```

File: Python/pystate.c
Comment (ericsnowcurrently) [inline Python/pystate.c]: FWIW, I liked having all the locks spelled out in one place via `LOCKS_INIT()`.  It made it much easier to add more locks.  However, this is fine since we only have the one place now where we mess with all the locks together.

File: Python/pystate.c
Comment (ericsnowcurrently) [inline Python/pystate.c:500]: I'm guessing we'll have another PR for per-interpreter locks, etc.

**Review by ericsnowcurrently — COMMENTED** [review]
mostly LGTM

I've left a few very minor comments.

File: Python/pylifecycle.c
Comment (colesbury) [reply Python/pylifecycle.c:3058]: `Py_AtExit` adds new at-exit handlers; it's called during normal operation. `call_ll_exitfuncs` below is called during shutdown.

We could use `_Py_LOCK_DONT_DETACH` here, but I don't think it matters. It seemed simpler to just use `PyMutex_Lock()` everywhere except in `HEAD_LOCK()` where we need `_Py_LOCK_DONT_DETACH`.

File: Python/ceval_gil.c
Comment (colesbury) [reply Python/ceval_gil.c]: I added `_PyMutex_at_fork_reinit`.

File: Python/pystate.c
Comment (colesbury) [reply Python/pystate.c]: I've re-added `LOCKS_INIT()`, but I hope this is the only place we need to enumerate them.

File: Python/pystate.c
Comment (colesbury) [reply Python/pystate.c:500]: Yeah, I think so.

File: Python/pylifecycle.c
Comment (ericsnowcurrently) [reply Python/pylifecycle.c:3058]: Yeah, that's what I meant.  I'm fine with leaving it.

File: Include/internal/pycore_lock.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_lock.h:99]: This is the same as `*m = (PyMutex}{0};`, right?  I'm fine with using `memset()`, though I find the initializer version a little easier to follow.  I'm not aware of the tradeoffs, so I'll leave it up to you.

File: Python/pystate.c
Comment (ericsnowcurrently) [reply Python/pystate.c]: Thanks!

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

I've left one comment that I'll leave to your discretion.  I'll wait for your response before merging.

File: Include/internal/pycore_lock.h
Comment (colesbury) [reply Include/internal/pycore_lock.h:99]: I originally wrote `*m = (PyMutex}{0};`, but that broke the Windows build. `*m = (PyMutex}{0};` is valid C,  but it is not valid C++. For some reason the Windows build compiles this header as part of a C++ file.

Comment (ericsnowcurrently) [reply Include/internal/pycore_lock.h:99]: blech!  okay then

**Comment by ericsnowcurrently** [comment]
Thanks for tackling this, @colesbury!

**Pull request merged**
